### PR TITLE
Fix Shortcut to toggle grid.

### PR DESCRIPTION
**Pull request by Carreau**
grid_size have been removed as part of #2144,

I'm not too sure how it was working, as shape seem to always be (-1, -1)
on the few examples I've tried; so replicate what the grid button does
seem to do

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

$ napari

Check that ctrl-G change the icon on the grid button.  (I'm not too sure what this is supposed to do...)

I'm not too sure how to test that; but with the work on the action manager we should have more confidence that the buttons and shortcut do actually trigger the same effects.

### CODE REVIEW

**Review by sofroniewn — APPROVED** [review]
Yes, this looks good, definitely an outdated API. I'm surprised this wasn't caught by any of our existing tests. Specifically it should have been caught here https://github.com/napari/napari/blob/5a2b3c31b5641cd27a44adde11a0ed377922f789/napari/_tests/test_viewer.py#L37-L44 can you see why it was missed?

**Comment by Carreau** [comment]
I'm not too sure why it did not fail; but I can investigate.

**Comment by sofroniewn** [comment]
> I'm not too sure why it did not fail; but I can investigate.

Thanks!

**Comment by Carreau** [comment]
viewer.class_keymap is en empty dict, the bindings are on viewer model, not viewer, so the for loop does nothing.

**Comment by Carreau** [comment]
So, move loop outside the test as a parametrized fixture as one of them was failing;  actual_shape ended up failing when nlayer = 0  

```
if n_row == -1 and n_column == -1:
                n_column = np.ceil(np.sqrt(n_grid_squares)).astype(int) # 0, 
                n_row = np.ceil(n_grid_squares / n_column).astype(int). # div by zero.
```

I'm not sure if that's right or if you want to actually load an empty layer in the viewer.

**Comment by Carreau** [comment]
in 
```
@pytest.mark.parametrize('layer_class, data, ndim', layer_test_data)
@pytest.mark.parametrize('visible', [True, False])
def test_add_layer(make_napari_viewer, layer_class, data, ndim, visible):
    viewer = make_napari_viewer()
    layer = add_layer_by_type(viewer, layer_class, data, visible=visible)
    check_viewer_functioning(viewer, viewer.window.qt_viewer, data, ndim)

    methods = set(super(type(layer), layer).class_keymap.values())
    methods.update(layer.class_keymap.values())
    # Run all class key bindings
    assert len(methods) in (14, 8, 17), layer
    for func in methods:
        func(layer)

```

the Image type layers has no functions; is that expected ? I don't see image specific buttons in the GUI; so that might be fine, but I want toget confirmation first; 

Should I update layer_test_data to contain the number of function we are expected to test ?

File: napari/_tests/test_viewer.py [hunk 1]
```diff
@@ -33,15 +41,12 @@ def test_viewer(make_napari_viewer):
...
-
     # Run all class key bindings
-    for func in viewer.class_keymap.values():
-        # skip fullscreen test locally
-        if func.__name__ == 'toggle_fullscreen' and not os.getenv("CI"):
-            continue
-        if func.__name__ == 'play':
-            continue
-        func(viewer)
+    if func.__name__ == 'toggle_fullscreen' and not os.getenv("CI"):
```
Comment (sofroniewn) [inline napari/_tests/test_viewer.py]: Can you pull this out into its own test, `tes_viewer_bindings`, that doesn't have any of the `assert` stuff above, no need for that to get tested each time

File: napari/_tests/test_viewer.py [hunk 2]
```diff
@@ -12,8 +12,16 @@
 )
 from napari.utils._tests.test_naming import eval_with_filename
 
+viewer_methods = set(super(Viewer, Viewer).class_keymap.values())
+viewer_methods.update(Viewer.class_keymap.values())
 
-def test_viewer(make_napari_viewer):
+
+def test_len_methods_viewer():
```
Comment (sofroniewn) [inline napari/_tests/test_viewer.py:26]: Can you add a comment doc string about what the `18` is? (I guess number of keybindings on the viewer).

File: napari/components/grid.py [hunk 3]
```diff
@@ -54,6 +54,8 @@ def actual_shape(self, nlayers: int = 1) -> Tuple[int, int]:
             Number of rows and columns in the grid.
         """
         if self.enabled:
+            if nlayers == 0:
```
Comment (sofroniewn) [inline napari/components/grid.py:57]: fine with me

**Review by sofroniewn — COMMENTED** [review]
> viewer.class_keymap is en empty dict, the bindings are on viewer model, not viewer, so the for loop does nothing.

Ah ok, clearly a bad test! nice catch. pulling this out into its own parametrized fixture makes sense, I think that could actually be its own test, separate from `test_viewer`, see comment above.

> the Image type layers has no functions; is that expected ? I don't see image specific buttons in the GUI; so that might be fine, but I want toget confirmation first;

That is right. I don't love `assert len(methods) in (14, 8, 17)` though, it feels a little magical and hard to remember where those numbers came from. Do you want to create a dict with the layer type and the number of expected methods and then check that it matches?

**Comment by Carreau** [comment]
> I don't love `assert len(methods) in (14, 8, 17)` though, it feels a little magical and hard to remember where those numbers came from. Do you want to create a dict with the layer type and the number of expected methods and then check that it matches?

Yes, that was the plan in the end. I just had to find all the layer types and which number for which layers :-) 

I've properly pulled the test into separate function to avoid costly computation; there seem to be other logic that is not exercised ; 'play' and 'toggle_fullscreen'  are not existing methods on viewer; I've added a test that I marked as xfailed.

I've also tweaked pytest configuration to ignore some warnings in dependencies; and set pytest to print the slowest 10 test by default if that's ok.

There is probably more refactoring that could be done, but this will likely be affected by the action manager in the end; soI'm not too sure it is worth getting into those refactor now.

**Comment by sofroniewn** [comment]
> Yes, that was the plan in the end. I just had to find all the layer types and which number for which layers :-)

Thanks!

> There is probably more refactoring that could be done, but this will likely be affected by the action manager in the end; soI'm not too sure it is worth getting into those refactor now.

Yeah, I think we should just get all the tests passing and merge.

It looks like we've got some sporadic fails in keybindings on our shapes layer. It doesn't seem to happen every time though. I'm not sure what's going on though - see

```
  >               if self._value[0] is None:
  E               TypeError: 'NoneType' object is not subscriptable
```

I'll try rerunning tests now. I wonder if it is significant that it fails on `3.9` and not `3.8`, my hunch is not, but we'll see!

**Comment by Carreau** [comment]
Hum it fails on 3.9 again; it's not the first project where I see things failing only on 3.9, and I don't know of anybody using 3.9 as their production environment; so it's starting to make me worried about 3.9 adoption...

**Comment by Carreau** [comment]
I can't reproduce locally so I'm going to attempt pushing some commits for debug. 
It _may_ be due "visibility" race condition that is the only way I see for this value to be set to `None`.

EDIT: (of course with debug statements test are passing...)

**Comment by Carreau** [comment]
Ok, I've parametrized the `add_layer`, and things seem to pass now. Maybe interaction when calling multiple methods on the same viewer ?

**Comment by sofroniewn** [comment]
> Ok, I've parametrized the add_layer, and things seem to pass now. Maybe interaction when calling multiple methods on the same viewer ?

Ok yeah, this is a much cleaner test, much better to see which keybinding the test fails on rather than having the for loop inside the test, thanks very much! A little bit concerning about the mystery fails, but it's fine! I will merge now :-)

**Comment by Carreau** [comment]
> Ok yeah, this is a much cleaner test, much better to see which keybinding the test fails on rather than having the for loop inside the test, thanks very much! A little bit concerning about the mystery fails, but it's fine! I will merge now :-)

Thanks ! the drawback is that the test-suite is a bit slower :-) Hopefully with the Action manager we'll have something even better.

**Pull request merged**